src/uct/ib/ud/base/ud_iface.h

```diff
@@ -381,24 +381,25 @@ uct_ud_iface_progress_pending(uct_ud_iface_t *iface, const uintptr_t is_async)
         return;
     }
 
-    iface->tx.in_pending = 1;
-    ucs_arbiter_dispatch(&iface->tx.pending_q, 1,
-             uct_ud_ep_do_pending, (void *)is_async);
-    iface->tx.in_pending = 0;
+    ucs_arbiter_dispatch(&iface->tx.pending_q, 1, uct_ud_ep_do_pending,
+                         (void *)is_async);
 }
 
-static UCS_F_ALWAYS_INLINE void
-uct_ud_iface_progress_pending_tx(uct_ud_iface_t *iface)
+static UCS_F_ALWAYS_INLINE int
+uct_ud_iface_had_async_pending(uct_ud_iface_t *iface)
 {
-    if (ucs_unlikely(iface->tx.pending_q_len > 0 &&
-                     iface->tx.in_pending == 0)) {
-        iface->tx.in_pending = 1;
-        ucs_arbiter_dispatch(&iface->tx.pending_q, 1,
-                             uct_ud_ep_do_pending, (void *)0);
-        iface->tx.in_pending = 0;
+    if (ucs_unlikely(iface->async.ev_count > 0)) {
+        return 1;
```

Comment:
maybe add assert that `ev_count` is not negative?
probably can just `return !iface->async.ev_count ` instead